File: web-service/api-project/src/main/java/io/ehdev/conrad/service/api/project/ProjectEndpoint.java
package io.ehdev.conrad.service.api.project;

import io.ehdev.conrad.database.api.ProjectManagementApi;
import io.ehdev.conrad.database.api.exception.ProjectAlreadyExistsException;
import io.ehdev.conrad.model.project.ApiProject;
import io.ehdev.conrad.model.user.ConradUser;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;

import javax.servlet.http.HttpServletRequest;
import javax.validation.Valid;
import javax.validation.constraints.NotNull;
import java.net.URI;

@Controller
@RequestMapping("/api/v1/project")
public class ProjectEndpoint {

    private final ProjectManagementApi projectManagementApi;

    @Autowired
    public ProjectEndpoint(ProjectManagementApi projectManagementApi) {
        this.projectManagementApi = projectManagementApi;
    }

    @RequestMapping(value = "/{projectName}", method = RequestMethod.POST)
    public ResponseEntity<ApiProject> createProject(@PathVariable("projectName") String projectName,
                                                    HttpServletRequest request,
                                                    @Valid @NotNull ConradUser user) {
        try {
            ApiProject project = projectManagementApi.createProject(projectName);
            return ResponseEntity.created(URI.create(request.getRequestURL().toString())).body(project);
        } catch (ProjectAlreadyExistsException e) {
            return new ResponseEntity<>(HttpStatus.CONFLICT);
        }
    }
}

File: internal/endpoints/src/main/kotlin/tech/crom/service/api/project/ProjectEndpoint.kt
package tech.crom.service.api.project

import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.MediaType
import org.springframework.http.ResponseEntity
import org.springframework.stereotype.Controller
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestMethod
import org.springframework.web.bind.annotation.RequestParam
import tech.crom.business.api.ProjectApi
import tech.crom.rest.model.project.FindProjectResponse

@Controller
@RequestMapping(value = ["/api/v1/projects"], produces = [MediaType.APPLICATION_JSON_VALUE])
class ProjectEndpoint @Autowired constructor(
    val projectApi: ProjectApi
) {

    @RequestMapping(method = [RequestMethod.GET])
    fun findProjects(@RequestParam("offset", defaultValue = "0") offset: String,
                     @RequestParam("size", defaultValue = "50") size: String): ResponseEntity<FindProjectResponse> {
        val offsetInt = offset.toInt()
        val sizeInt = size.toInt()

        val projects = projectApi.findProjects(offsetInt, sizeInt)
        val projectDetails = projects.projects.map { FindProjectResponse.ProjectDetails(it.projectName) }

        return ResponseEntity.ok(FindProjectResponse(projectDetails, projects.totalSize))
    }

}
